[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Nia-04/CSPC-513-ML-Lab/blob/main/Practical_03_KNN.ipynb)

# Practical No. 3: Implementation of K-Nearest Neighbors (KNN)

## Aim
To implement the K-Nearest Neighbors (KNN) classification algorithm on the processed dataset using different distance measures and evaluate its performance.

## Introduction

K-Nearest Neighbors (KNN) is a supervised machine learning algorithm used for both classification and regression problems. It is a simple, instance-based learning algorithm that makes predictions based on the similarity between data points.

In KNN classification, the class of a new data point is determined by examining its K nearest data points in the training dataset. The class that occurs most frequently among the nearest neighbors is assigned to the new data point.

In this practical, KNN is applied to the processed student dataset. The normalized marks of five subjects are used as input features, while the student's grade is used as the target variable.

In [28]:
import os

if os.path.exists("EXPORTS/Mldata_Preprocessed.xlsx"):
    print("Dataset folder already available.")
else:
    !git clone https://github.com/Nia-04/CSPC-513-ML-Lab.git
    %cd CSPC-513-ML-Lab
    print("Repository cloned successfully.")

Dataset folder already available.


In [1]:
import pandas as pd
import numpy as np

## Dataset Used

The processed dataset obtained from Practical 2 is used as the input dataset.

The five normalized subject marks used as input features (`X`) are:

- Software Engineer
- Theory of Computation (TOC)
- Computer Network
- Machine Learning
- Data Warehouse

The `GRADE` column is used as the target variable (`y`).

The dataset contains **67 student records** and **5 input features**.

In [2]:
df = pd.read_excel("EXPORTS/Mldata_Preprocessed.xlsx")
df.head()

,ROLL NO.,GENDER,SOFTWARE ENGINEER,TOC,COMPUTER NETWORK,MACHINE LEARNING,DATA WAREHOUSE,GRADE,RESULT
0,240104010001,M,0.500000,0.491525,0.482759,0.473684,0.454545,C,1
1,240104010002,F,0.500000,0.661017,0.741379,0.649123,0.545455,B,1
2,240104010003,M,0.633333,0.627119,0.620690,0.614035,0.600000,B,1
3,240104010004,M,0.983333,0.915254,0.948276,0.964912,0.890909,O,1
4,240104010005,M,0.500000,0.491525,0.482759,0.473684,0.454545,C,1


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ROLL NO.           67 non-null     int64  
 1   GENDER             67 non-null     object 
 2   SOFTWARE ENGINEER  67 non-null     float64
 3   TOC                67 non-null     float64
 4   COMPUTER NETWORK   67 non-null     float64
 5   MACHINE LEARNING   67 non-null     float64
 6   DATA WAREHOUSE     67 non-null     float64
 7   GRADE              67 non-null     object 
 8   RESULT             67 non-null     int64  
dtypes: float64(5), int64(2), object(2)
memory usage: 4.8+ KB


In [4]:
df['GRADE'].value_counts()

GRADE
O     18
C     13
B     12
A     10
A+     7
P      4
F      3
Name: count, dtype: int64

### K-Nearest Neighbors (KNN)

K-Nearest Neighbors is a supervised, non-parametric machine learning algorithm. It does not build an explicit mathematical model during training. Instead, it stores the training data and makes predictions by finding the nearest data points to a given test point.

The basic steps of KNN are:

1. Choose the value of K.
2. Calculate the distance between the test point and all training points.
3. Identify the K nearest neighbors.
4. Determine the most common class among these neighbors.
5. Assign that class to the test point.

In this practical, **K = 5** is used for all three distance measures to ensure a fair comparison.

In [7]:
X = df[['SOFTWARE ENGINEER',
        'TOC',
        'COMPUTER NETWORK',
        'MACHINE LEARNING',
        'DATA WAREHOUSE']]

In [9]:
y = df['GRADE']

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [14]:
print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

print("\nTraining target distribution:")
print(y_train.value_counts())

print("\nTesting target distribution:")
print(y_test.value_counts())

Training data: (53, 5)
Testing data: (14, 5)

Training target distribution:
GRADE
O     14
C     10
B     10
A      8
A+     6
P      3
F      2
Name: count, dtype: int64

Testing target distribution:
GRADE
O     4
C     3
A     2
B     2
A+    1
P     1
F     1
Name: count, dtype: int64


In [15]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## 1. Euclidean Distance

Euclidean distance measures the straight-line distance between two points. It is commonly used when the distance between data points is measured in a geometric space.

**Formula:**

$$
d(x,y) =
\sqrt{
\sum_{i=1}^{n}(x_i-y_i)^2
}
$$

In Scikit-learn:

`metric='euclidean'`


In [16]:
knn_euclidean = KNeighborsClassifier(
    n_neighbors=5,
    metric='euclidean'
)
knn_euclidean.fit(X_train, y_train)
y_pred_euclidean = knn_euclidean.predict(X_test)

In [17]:
accuracy_euclidean = accuracy_score(y_test, y_pred_euclidean)
print("Euclidean Distance Accuracy:",
      accuracy_euclidean * 100, "%")

Euclidean Distance Accuracy: 64.28571428571429 %


### Confusion Matrix

A confusion matrix is a table used to evaluate the classification performance of a model. It shows the number of correctly and incorrectly classified samples for each class.

The rows represent the actual classes, while the columns represent the predicted classes.

### Classification Report

The classification report provides:

- **Precision** – proportion of predicted samples that are actually correct.
- **Recall** – proportion of actual samples correctly identified.
- **F1-score** – harmonic mean of precision and recall.
- **Support** – number of actual samples belonging to each class.

In [19]:
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_euclidean))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_euclidean, zero_division=0))


Confusion Matrix:
[[0 0 2 0 0 0 0]
 [0 0 0 0 0 1 0]
 [0 0 2 0 0 0 0]
 [0 0 0 3 0 0 0]
 [0 0 0 1 0 0 0]
 [0 0 0 0 0 4 0]
 [0 0 0 1 0 0 0]]

Classification Report:
              precision    recall  f1-score   support

           A       0.00      0.00      0.00         2
          A+       0.00      0.00      0.00         1
           B       0.50      1.00      0.67         2
           C       0.60      1.00      0.75         3
           F       0.00      0.00      0.00         1
           O       0.80      1.00      0.89         4
           P       0.00      0.00      0.00         1

    accuracy                           0.64        14
   macro avg       0.27      0.43      0.33        14
weighted avg       0.43      0.64      0.51        14



## 2. Manhattan Distance

Manhattan distance calculates the distance between two points by adding the absolute differences between their corresponding features.

**Formula:**

$$
d(x,y) =
\sum_{i=1}^{n}|x_i-y_i|
$$

In Scikit-learn:

`metric='manhattan'`


In [20]:
knn_manhattan = KNeighborsClassifier(
    n_neighbors=5,
    metric='manhattan'
)
knn_manhattan.fit(X_train, y_train)
y_pred_manhattan = knn_manhattan.predict(X_test)

In [21]:
accuracy_manhattan = accuracy_score(y_test, y_pred_manhattan)
print("Manhattan Distance Accuracy:",
      accuracy_manhattan * 100, "%")

Manhattan Distance Accuracy: 78.57142857142857 %


In [22]:
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_manhattan))
print("\nClassification Report:")
print(classification_report(
    y_test,
    y_pred_manhattan,
    zero_division=0
))


Confusion Matrix:
[[1 0 1 0 0 0 0]
 [0 0 0 0 0 1 0]
 [0 0 2 0 0 0 0]
 [0 0 0 3 0 0 0]
 [0 0 0 0 0 0 1]
 [0 0 0 0 0 4 0]
 [0 0 0 0 0 0 1]]

Classification Report:
              precision    recall  f1-score   support

           A       1.00      0.50      0.67         2
          A+       0.00      0.00      0.00         1
           B       0.67      1.00      0.80         2
           C       1.00      1.00      1.00         3
           F       0.00      0.00      0.00         1
           O       0.80      1.00      0.89         4
           P       0.50      1.00      0.67         1

    accuracy                           0.79        14
   macro avg       0.57      0.64      0.57        14
weighted avg       0.72      0.79      0.73        14



## 3. Minkowski Distance

Minkowski distance is a generalized distance measure that can represent different distance metrics depending on the value of \(p\).

**Formula:**

$$
d(x,y) =
\left(
\sum_{i=1}^{n}|x_i-y_i|^p
\right)^{1/p}
$$

The value of \(p\) determines the type of distance:

* \(p = 3\) → Minkowski distance used in this practical

In Scikit-learn:

`metric='minkowski', p=3`


In [23]:
knn_minkowski = KNeighborsClassifier(
    n_neighbors=5,
    metric='minkowski',
    p=3
)
knn_minkowski.fit(X_train, y_train)
y_pred_minkowski = knn_minkowski.predict(X_test)

In [24]:
accuracy_minkowski = accuracy_score(y_test, y_pred_minkowski)
print("Minkowski Distance Accuracy:",
      accuracy_minkowski * 100, "%")

Minkowski Distance Accuracy: 71.42857142857143 %


In [25]:
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_minkowski))
print("\nClassification Report:")
print(classification_report(
    y_test,
    y_pred_minkowski,
    zero_division=0
))


Confusion Matrix:
[[0 0 2 0 0 0 0]
 [0 1 0 0 0 0 0]
 [0 0 2 0 0 0 0]
 [0 0 0 3 0 0 0]
 [0 0 0 1 0 0 0]
 [0 0 0 0 0 4 0]
 [0 0 0 1 0 0 0]]

Classification Report:
              precision    recall  f1-score   support

           A       0.00      0.00      0.00         2
          A+       1.00      1.00      1.00         1
           B       0.50      1.00      0.67         2
           C       0.60      1.00      0.75         3
           F       0.00      0.00      0.00         1
           O       1.00      1.00      1.00         4
           P       0.00      0.00      0.00         1

    accuracy                           0.71        14
   macro avg       0.44      0.57      0.49        14
weighted avg       0.56      0.71      0.61        14



### Accuracy

Accuracy represents the proportion of correctly classified samples out of all test samples.

$$
Accuracy =
\frac{Correct\ Predictions}
{Total\ Predictions}
$$


In [26]:
results = pd.DataFrame({
    'Distance': ['Euclidean', 'Manhattan', 'Minkowski'],
    'Accuracy': [
        accuracy_euclidean,
        accuracy_manhattan,
        accuracy_minkowski
    ]
})
results['Accuracy (%)'] = results['Accuracy'] * 100
results

,Distance,Accuracy,Accuracy (%)
0,Euclidean,0.642857,64.285714
1,Manhattan,0.785714,78.571429
2,Minkowski,0.714286,71.428571


In [27]:
best_distance = results.loc[
    results['Accuracy'].idxmax(),
    'Distance'
]
best_accuracy = results['Accuracy (%)'].max()
print("Best Distance:", best_distance)
print("Best Accuracy:", best_accuracy, "%")

Best Distance: Manhattan
Best Accuracy: 78.57142857142857 %


### Conclusion

K-Nearest Neighbors (KNN) was successfully implemented on the processed student dataset using three different distance measures: Euclidean, Manhattan, and Minkowski distance. The five normalized subject marks were used as input features and `GRADE` was used as the target variable.

The obtained accuracies were 64.29% for Euclidean distance, 78.57% for Manhattan distance, and 71.43% for Minkowski distance with p = 3. Among the three distance measures, **Manhattan distance performed the best for this dataset**, achieving an accuracy of **78.57%**.

The confusion matrices and classification reports showed that the model performed better for classes with more samples, while performance for classes with very few samples such as F, P, and A+ was less reliable. This is mainly due to the small size of the dataset and the uneven distribution of grades.